text/0162-refactoring-support.md

```diff
@@ -0,0 +1,586 @@
+been captured earlier, encoded as a mapping between resource locations, and
+stored somewhere.
+
+But this is not sufficient. Note that every mapping is created from a pair of
+source and target states, out of which the ambiguities arose. To be able to
+safely carry a mapping over to other environments, two additional conditions
+must be met:
+
+1. The source state on which a mapping is applied must be the same as the source
+   state where the mapping was captured.
+2. The target state used to create the mapping should indeed be what the user
+   wants as a result.
+
+I am using the abstract term "state" here, but how could such a state be
+instantiated in practice? Let's consider some options and see how they fail to
+satisfy the conditions above.
+
+First, we need to establish a point when the mapping is created (and the
+developer is involved to resolve possible ambiguities). Let's call this the
+"decision point". As a first attempt, let's try to use every deployment in the
+development cycle as the decision point. In this solution, the development
+account is the source state, and the cloud assembly to be deployed is the target
+state. If any ambiguities were resolved, they are saved in a mapping file, under
+version control. On every deployment to other environments, the mapping file is
+used to perform the refactoring.
+
+It sounds like this could work, but if the development environment is not in the
+same state as the one where the mapping is applied, condition 1 is violated. And
+if the developer fails, for whatever reason, to run a deployment against their
+environment before commiting an ambiguous change to the version control system
+(which I will henceforth assume is Git), condition 2 is violated.
+
+Since we are talking about Git, what about using each commit operation as a
+decision point? In this case, the source and target states would come from the
+synthesized cloud assemblies in the previous and current revision, respectively.
+We still have a mapping file, containing ambiguity resolutions, which are added
+to the commit, using a Git hook. For this solution to work, we need an
+additional constraint, which can also be enforced with a Git hook: that every
+revision produces a valid cloud assembly.
+
+Let's evaluate this solution in terms of the two conditions above. Because the
+developer doesn't have a choice anymore of which target state to use (or source
+state, for that matter), condition 2 is satisfied. But remember that the scope
+of the mapping file is the difference between two consecutive revisions. If the
+developer's local branch is multiple commits ahead of the revision that was
+deployed to production, the source state in production is not the same as the
+one in the mapping file, violating condition 1.
+
+#### Making history
+
+An improvement we can make is to turn this into an event sourcing system.
+Instead of storing a single mapping between two states, we store the whole
```

Comment:
i assume the whole history needs to be stored in a persistent layer (S3?) that all parties using cdk can access?

Reply:
Yes, the history would have to be stored in two places: in Git, and in a persistent on AWS. S3 seems the most straightforward choice, but it could also be in DynamoDB, etc.